marker_api/src/diagnostic.rs

```diff
@@ -1,36 +1,74 @@
 //! This module is responsible for the construction of diagnostic messages. The
 //! [`DiagnosticBuilder`] is the public stable interface, to construct messages.
 
+use std::fmt::Debug;
+
 use crate::{
-    ast::{ExprId, FieldId, ItemId, Span, StmtId, VariantId},
-    context::AstContext,
+    ast::{Identifiable, NodeId, Span},
+    context::{with_cx, AstContext},
     ffi::{FfiSlice, FfiStr},
     lint::Lint,
+    prelude::Spanned,
 };
 
 /// This builder creates the diagnostic object which will be emitted by the driver.
 /// The documentation will showcase the messages in rustc's console emission style,
 /// the actual display depends on the driver.
 pub struct DiagnosticBuilder<'ast> {
+    /// This field will be `Some` if the created diagnostic will be emitted, otherwise
+    /// it'll be `None`. See [`DiagnosticBuilder::decorate`] for more information, when the
+    /// lint might be suppressed.
+    inner: Option<DiagnosticBuilderInner<'ast>>,
+}
+
+struct DiagnosticBuilderInner<'ast> {
     lint: &'static Lint,
+    node: NodeId,
     msg: String,
-    node: EmissionNode,
     span: Span<'ast>,
     parts: Vec<DiagnosticPart<String, Span<'ast>>>,
 }
 
-#[allow(clippy::needless_pass_by_value)] // `&impl ToString` doesn't work
+#[allow(clippy::needless_pass_by_value)] // `&impl Into<String>` doesn't work
 impl<'ast> DiagnosticBuilder<'ast> {
-    pub(crate) fn new(lint: &'static Lint, node: EmissionNode, msg: String, span: Span<'ast>) -> Self {
+    /// Creates a new dummy builder, which basically makes all operations a noop
+    pub(crate) fn dummy() -> Self {
+        Self { inner: None }
+    }
+
+    pub(crate) fn new(lint: &'static Lint, node: NodeId, msg: String, span: Span<'ast>) -> Self {
         Self {
-            lint,
-            msg,
-            node,
-            span,
-            parts: vec![],
+            inner: Some(DiagnosticBuilderInner {
+                lint,
+                msg,
+                node,
+                span,
+                parts: vec![],
+            }),
         }
     }
 
+    /// This function sets the main [`Span`] of the created diagnostic.
+    /// [`AstContext::emit_lint`] will by default use the [`Span`] of the given
+    /// [`EmissionNode`].
+    ///
+    /// From rustc a lint emission would look like this:
+    /// ```text
+    ///  warning: <lint message>
+    ///  --> path/file.rs:1:1
+    ///   |
+    /// 1 | node
+    ///   | ^^^^                 <-- The main span set by this function
+    ///   |
+    /// ```
+    pub fn set_span(&mut self, span: &Span<'ast>) -> &mut Self {
```

Comment:
Maybe this should be called just `span` (stripping the `set_` prefix)? Other methods like `note()` don't have `set_` prefix. 
For consistency it makes sense to either have either of the two:
- No `set_` prefix for setters. If getters are needed give them a `get_` prefix, like in `std::process::Command` API there is [`args()`](https://doc.rust-lang.org/stable/std/process/struct.Command.html#method.args) setter and [`get_args()`](https://doc.rust-lang.org/stable/std/process/struct.Command.html#method.get_args) getter.
- Add `set_` prefix for setters. If getters are needed give them no prefix.

The first approach looks better for builders where the bulk of the API usage is setters, and getters are almost never used or even non-existent.

Reply:
I'm not sure if `span()` might be too generic and also too similar to the `span()` method from all other nodes. I would also argue that `.note()` is slightly different in the way that it adds a note and doesn't override an existing one :thinking:.

The builder example speaks again for just calling it `span()`. How strong do you feel about either option?

Reply:
Rather strong. I believe the builder should follow `std::process::Command` builder design. That type has, for example `.current_dir()` method that is a setter which if you call it twice, the second call will override the first one's value.

But the methods `.arg()`, `.args()` don't override the previous values, they just push new ones. And like this API, personally. It's quite obvious and intuitive for me. I'm glad that guys from `std` didn't add any extra prefixes/suffixes.

Reply:
Okay, I don't have a strong opinion, so I'll take your suggestion :)